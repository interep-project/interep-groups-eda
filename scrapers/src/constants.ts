import { join } from 'path'

export const SAMPLE_SIZE = 1000
export const DATA_DIR = join(__dirname, '..', '..', 'data')

// found by manual binary search as of 3/2/23, assuming ID is an integer increased by 1 upon each new user registration
export const GH_USER_MAX_ID = 124_369_989
