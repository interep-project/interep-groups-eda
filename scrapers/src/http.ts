import wretch from "wretch";

export const http = (url: string) => wretch(url).resolve(async r => r.json())