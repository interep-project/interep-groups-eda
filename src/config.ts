import { existsSync, readFileSync } from 'fs'
import { join } from 'path'

import YAML from 'yaml'
import { Config } from 'types/config'

const path = join(__dirname, '..')
const exists = existsSync(join(path, '.config.yaml'))
const file = exists
  ? join(path, '.config.yaml')
  : join(path, '.config.test.yaml')

export const config: Config = YAML.parse(readFileSync(file, 'utf8'))
