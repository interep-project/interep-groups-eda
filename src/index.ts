import { Twitter } from './Twitter'

const main = async () => {
  const twitter = new Twitter()

  await twitter.loadUsers()
  await twitter.writeUsers()
}

main()
  .then(() => {
    process.exit(0)
  })
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
