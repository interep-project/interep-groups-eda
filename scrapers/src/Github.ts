import { faker } from '@faker-js/faker'
import { graphql } from '@octokit/graphql'
import { paginateGraphql } from '@octokit/plugin-paginate-graphql'
import { ok } from 'assert'
// import { writeFile } from 'fs/promises'

import { Octokit } from 'octokit'
import { GH_USER_MAX_ID } from './constants'
import { Provider } from './Provider'
import { getUserQuery } from './queries/get-gh-users.gql'

// TODO define GH type
export class Github extends Provider<any> {
  client: Octokit & {
    graphql: typeof graphql & {
      paginate: (
        query: string,
        initialParameters?: Record<string, any> | undefined,
      ) => Promise<any>
    }
  }

  constructor() {
    super('gh')
    ok(process.env.GH_PAT, 'GH_PAT is not defined')
    const PaginatedOctokit = Octokit.plugin(paginateGraphql)
    this.client = new PaginatedOctokit({ auth: process.env.GH_PAT })
  }

  randomIds(): string[] {
    const max = Math.floor((GH_USER_MAX_ID - 100) / 100)
    const startId = faker.datatype.number({ max, min: 1 })
    return this.usedIds.includes(startId.toString())
      ? this.randomIds()
      : [startId.toString()]
  }

  async fetchUserNodeIds(): Promise<string[]> {
    const [since] = this.randomIds()
    const users = await this.client.rest.users.list({
      per_page: 100,
      since: Number(since),
    })
    // console.log(since, users)
    return users.data.map((user) => user.node_id)
  }

  async fetchUsers(): Promise<
    Record<
      string,
      {
        followers: number
        issues: number
        login: string
        pullRequests: number
        sponsoring: number
        sponsors: number
        stars: number
        forks: number
        totalContributions: number
      }
    >
  > {
    const ids = await this.fetchUserNodeIds()
    let data: any[]

    try {
      // @ts-expect-error
      data = (await this.client.graphql(getUserQuery, { ids }))?.nodes
      // console.log(data)
    } catch (error) {
      // console.log(error);
      // @ts-expect-error
      data = error.data?.nodes
    }

    return data === undefined
      ? {}
      : data.reduce<
          Record<
            string,
            {
              followers: number
              issues: number
              login: string
              pullRequests: number
              sponsoring: number
              sponsors: number
              totalContributions: number
              forks: number
              stars: number
            }
          >
        >((users, user) => {
          if (user === null) return users
          const followers = user?.followers?.totalCount
          const issues = user?.issues?.totalCount
          const login = user?.login
          const pullRequests = user.pullRequests?.totalCount
          const sponsors = user?.sponsors?.totalCount
          const sponsoring = user?.sponsoring?.totalCount
          const totalContributions =
            user?.contributionsCollection?.contributionCalendar
              ?.totalContributions

          const { forks, stars } = (
            (user?.repositories?.nodes as Array<{
              stargazerCount: number
              forkCount: number
            }>) ?? []
          ).reduce(
            (acc, { forkCount, stargazerCount }) => ({
              forks: acc.forks + forkCount,
              stars: acc.stars + stargazerCount,
            }),
            {
              forks: 0,
              stars: 0,
            },
          )

          if (
            ![
              login,
              followers,
              totalContributions,
              issues,
              pullRequests,
              sponsors,
              sponsoring,
            ].includes(undefined)
          )
            users[login] = {
              followers,
              forks,
              issues,
              login,
              pullRequests,
              sponsoring,
              sponsors,
              stars,
              totalContributions,
            }

          return users
        }, {})
  }
}

export const github = new Github()
