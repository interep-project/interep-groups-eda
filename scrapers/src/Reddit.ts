import FormUrlAddon from 'wretch/addons/formUrl'
import { config } from './config'
import { http } from './http'

import { Provider } from './Provider'

export class Reddit extends Provider<any> {
  private token: string | undefined

  constructor() {
    super('reddit')
  }

  private async auth(): Promise<void> {
    try {
      const response = await http('https://www.reddit.com/api/v1/access_token')
        .addon(FormUrlAddon)
        .formUrl({
          device_id: 'DO_NOT_TRACK_THIS_DEVICE',
          grant_type: 'https://oauth.reddit.com/grants/installed_client',
        })
        .auth(
          `Basic ${Buffer.from(
            `${config.reddit.clientId}:${config.reddit.clientSecret}`,
          ).toString('base64')}`,
        )
        .post()

      // @ts-expect-error
      if (response.error === undefined) {
        // @ts-expect-error
        this.token = response.access_token
      }
    } catch (err) {
      console.log(err)
    }
  }

  private async authed_get_request(
    path: string,
    query?: Record<string, any>,
  ): Promise<any> {
    if (this.token === undefined) {
      await this.auth()
    }

    return (
      http(
        `https://oauth.reddit.com/${path}?${new URLSearchParams(
          query,
        ).toString()}`,
      )
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
        .auth(`Bearer ${this.token}`)
        .get()
    )
  }

  async randomAuthors(): Promise<any> {
    try {
      const json = await this.get_random_subreddit()
      return (json?.data?.children ?? []).map(
        (child: { data: { author: string } }) => child.data.author,
      )
    } catch (err) {
      console.log(err)
      return []
    }
  }

  private async fetchUser(username: string): Promise<any> {
    const user = await this.authed_get_request(`user/${username}/about`)
    return { [user.data.name]: user.data }
  }

  async fetchUsers(): Promise<any[]> {
    const authors = await this.randomAuthors()
    const users = await Promise.all(authors.map(this.fetchUser.bind(this)))
    return users.reduce((acc, user) => ({ ...acc, ...user }), {})
  }

  private async get_random_subreddit(): Promise<any> {
    return http('https://www.reddit.com/r/random.json').get()
  }
}

export const reddit = new Reddit()
