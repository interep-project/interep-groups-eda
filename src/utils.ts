import { faker } from '@faker-js/faker'
import { Botometer as _Botometer } from 'botometer'

import { config } from './config'

export const randomTwitterIds = ({
  size = 100,
  usedIds = [],
}: {
  size?: number
  usedIds?: string[]
} = {}) => {
  const ids: string[] = []

  while (ids.length < size) {
    const id = faker.random.numeric(faker.datatype.number({ max: 19, min: 1 }))

    if (!usedIds.includes(id) && !ids.includes(id) && id !== '') {
      ids.push(id)
    }
  }

  return ids
}

class Botometer extends _Botometer {
  constructor() {
    const { bearerToken, ...rest } = config.twitter
    super({ ...rest, usePro: true })
  }

  async getScores(usernames: string[]) {
    console.log(usernames)
    const scores = await super.getScores(usernames)
    console.log(scores)
    return scores.reduce(
      (
        scores: { [x: string]: any },
        score: { [x: string]: any; error?: any; user?: any; twitterData?: any },
      ) => {
        if (score?.user?.user_data === undefined) return scores

        const { twitterData, ...scoreData } = score
        scores[score.user.user_data.id_str] = scoreData
        console.log(score.user.user_data.screen_name, score.user.user_data.id)

        return scores
      },
      {},
    )
  }
}

export const botometer = new Botometer()
