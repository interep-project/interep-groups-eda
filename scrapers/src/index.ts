import { existsSync, mkdirSync } from 'fs'

import { DATA_DIR } from './constants'
// import { twitter } from './Twitter'
// import { github } from './Github'
import { reddit } from './Reddit'

const maybeCreateDir = () => {
  if (!existsSync(DATA_DIR)) mkdirSync(DATA_DIR)
}

const main = async () => {
  maybeCreateDir()

  let sampleSize
  if (process.argv[2] !== undefined) sampleSize = Number(process.argv[2])
  await reddit.maybeWriteUsers(sampleSize)

  // await github.maybeWriteUsers()
}

main()
  .then(() => {
    process.exit(0)
  })
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
