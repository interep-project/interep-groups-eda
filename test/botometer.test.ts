import { botometer } from '../src/utils'

describe.skip('botometer', () => {
  it('fetches botometer scores for a list of usernames', async () => {
    const usernames = ['@r1oga', '@VitalikButerin']
    const scores = await botometer.getScores(usernames)

    expect(scores[0]).toMatchSnapshot()
    expect(scores[1]).toMatchSnapshot()
  })
})
