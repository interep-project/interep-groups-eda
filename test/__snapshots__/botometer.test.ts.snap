// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`botometer fetches botometer scores for a list of usernames 1`] = `
{
  "cap": {
    "english": 0.3342309004634953,
    "universal": 0.3469210182245441,
  },
  "display_scores": {
    "english": {
      "astroturf": 0.2,
      "fake_follower": 0,
      "financial": 0,
      "other": 0.4,
      "overall": 0.2,
      "self_declared": 0,
      "spammer": 0,
    },
    "universal": {
      "astroturf": 0.4,
      "fake_follower": 0,
      "financial": 0,
      "other": 0.5,
      "overall": 0.2,
      "self_declared": 0,
      "spammer": 0.1,
    },
  },
  "raw_scores": {
    "english": {
      "astroturf": 0.03,
      "fake_follower": 0.01,
      "financial": 0.01,
      "other": 0.08,
      "overall": 0.05,
      "self_declared": 0.01,
      "spammer": 0,
    },
    "universal": {
      "astroturf": 0.07,
      "fake_follower": 0.01,
      "financial": 0.01,
      "other": 0.1,
      "overall": 0.05,
      "self_declared": 0,
      "spammer": 0.02,
    },
  },
  "user": {
    "majority_lang": "en",
    "user_data": {
      "contributors_enabled": false,
      "created_at": "Mon Feb 19 23:06:16 +0000 2018",
      "default_profile": true,
      "default_profile_image": false,
      "description": "Web3 Engineer @ethereum @PrivacyScaling.      Prev: @exodus_io",
      "entities": {
        "description": {
          "urls": [],
        },
        "url": {
          "urls": [
            {
              "display_url": "github.com/r1oga",
              "expanded_url": "https://github.com/r1oga",
              "indices": [
                0,
                23,
              ],
              "url": "https://t.co/675lqgxQmM",
            },
          ],
        },
      },
      "favourites_count": 1968,
      "follow_request_sent": false,
      "followers_count": 57,
      "following": false,
      "friends_count": 20,
      "geo_enabled": false,
      "has_extended_profile": false,
      "id": 965724241726443500,
      "id_str": "965724241726443520",
      "is_translation_enabled": false,
      "is_translator": false,
      "lang": null,
      "listed_count": 1,
      "location": "üåê",
      "name": "r1oga.eth | ü¶áüîä",
      "notifications": false,
      "profile_background_color": "F5F8FA",
      "profile_background_image_url": null,
      "profile_background_image_url_https": null,
      "profile_background_tile": false,
      "profile_banner_url": "https://pbs.twimg.com/profile_banners/965724241726443520/1618155900",
      "profile_image_url": "http://pbs.twimg.com/profile_images/1447288032688750599/uWSsm2zX_normal.jpg",
      "profile_image_url_https": "https://pbs.twimg.com/profile_images/1447288032688750599/uWSsm2zX_normal.jpg",
      "profile_link_color": "1DA1F2",
      "profile_sidebar_border_color": "C0DEED",
      "profile_sidebar_fill_color": "DDEEF6",
      "profile_text_color": "333333",
      "profile_use_background_image": true,
      "protected": false,
      "screen_name": "r1oga",
      "statuses_count": 964,
      "time_zone": null,
      "translator_type": "none",
      "url": "https://t.co/675lqgxQmM",
      "utc_offset": null,
      "verified": false,
      "withheld_in_countries": [],
    },
  },
}
`;

exports[`botometer fetches botometer scores for a list of usernames 2`] = `
{
  "cap": {
    "english": 0.7053377817049258,
    "universal": 0.8030212452554465,
  },
  "display_scores": {
    "english": {
      "astroturf": 0.3,
      "fake_follower": 0.5,
      "financial": 0,
      "other": 1.6,
      "overall": 1,
      "self_declared": 0.2,
      "spammer": 0.2,
    },
    "universal": {
      "astroturf": 0.6,
      "fake_follower": 0.4,
      "financial": 0,
      "other": 2.8,
      "overall": 2,
      "self_declared": 0,
      "spammer": 0.1,
    },
  },
  "raw_scores": {
    "english": {
      "astroturf": 0.06,
      "fake_follower": 0.1,
      "financial": 0.01,
      "other": 0.33,
      "overall": 0.21,
      "self_declared": 0.04,
      "spammer": 0.03,
    },
    "universal": {
      "astroturf": 0.13,
      "fake_follower": 0.08,
      "financial": 0.01,
      "other": 0.57,
      "overall": 0.4,
      "self_declared": 0.01,
      "spammer": 0.02,
    },
  },
  "user": {
    "majority_lang": "en",
    "user_data": {
      "contributors_enabled": false,
      "created_at": "Sun May 08 16:03:03 +0000 2011",
      "default_profile": true,
      "default_profile_image": false,
      "description": "Ethereum

Fable of the Dragon Tyrant (not mine but it's important): https://t.co/t3jaMRRrPs

Abolish daylight savings time and leap seconds",
      "entities": {
        "description": {
          "urls": [
            {
              "display_url": "youtube.com/watch?v=cZYNAD‚Ä¶",
              "expanded_url": "https://www.youtube.com/watch?v=cZYNADOHhVY",
              "indices": [
                68,
                91,
              ],
              "url": "https://t.co/t3jaMRRrPs",
            },
          ],
        },
        "url": {
          "urls": [
            {
              "display_url": "vitalik.ca",
              "expanded_url": "http://vitalik.ca",
              "indices": [
                0,
                23,
              ],
              "url": "https://t.co/IHF9ic0Ejh",
            },
          ],
        },
      },
      "favourites_count": 6741,
      "follow_request_sent": false,
      "followers_count": 4704830,
      "following": true,
      "friends_count": 351,
      "geo_enabled": false,
      "has_extended_profile": true,
      "id": 295218901,
      "id_str": "295218901",
      "is_translation_enabled": false,
      "is_translator": false,
      "lang": null,
      "listed_count": 31164,
      "location": "Earth",
      "name": "vitalik.eth",
      "notifications": false,
      "profile_background_color": "C0DEED",
      "profile_background_image_url": "http://abs.twimg.com/images/themes/theme1/bg.png",
      "profile_background_image_url_https": "https://abs.twimg.com/images/themes/theme1/bg.png",
      "profile_background_tile": false,
      "profile_banner_url": "https://pbs.twimg.com/profile_banners/295218901/1638557376",
      "profile_image_url": "http://pbs.twimg.com/profile_images/977496875887558661/L86xyLF4_normal.jpg",
      "profile_image_url_https": "https://pbs.twimg.com/profile_images/977496875887558661/L86xyLF4_normal.jpg",
      "profile_link_color": "1DA1F2",
      "profile_sidebar_border_color": "C0DEED",
      "profile_sidebar_fill_color": "DDEEF6",
      "profile_text_color": "333333",
      "profile_use_background_image": true,
      "protected": false,
      "screen_name": "VitalikButerin",
      "statuses_count": 19252,
      "time_zone": null,
      "translator_type": "none",
      "url": "https://t.co/IHF9ic0Ejh",
      "utc_offset": null,
      "verified": true,
      "withheld_in_countries": [],
    },
  },
}
`;
